### DISCUSSION TITLE
Speed up MySQL information_schema read

### DISCUSSION THREAD

**Discussion by jsthind** [directus/directus#5305] [Feature Requests]
I did a fresh installation using on my computer (running Ubuntu 20)

`npx create-directus-project my-project`

Selected MySQL for the database. The MySQL server is running on an HP1 Proliant server on the LAN. Tried to login, the login seems to work fine but the http://127.0.0.1:8055/admin/collections page times out. 

See screenshot:

![directus-error](https://user-images.githubusercontent.com/456605/101148551-c45bee80-3643-11eb-91e5-1aa5c2a779bc.png)

**Comment by rijkvanzanten** [comment]
That `ERR_EMPTY_RESPONSE` feels suspicious to me. What kind of response are you seeing there? What webserver are you running locally to serve this? Just Node?

**Comment by jsthind** [comment]
I am running on Node. No other proxy or server. 

The /fields endpoint keeps waiting for a response and then times out with an ERR_EMPTY_RESPONSE error.

**Comment by rijkvanzanten** [comment]
But you are able to login? It feels like the query to the DB to extract the fields information times out 🤔 How are you connecting to MySQL? Does your MySQL user have a restricted permissions set?

**Comment by jsthind** [comment]
I have been using the same MySQL installation with the same credentials (username and password) for years without any issue. It is a standard MySQL installation on Ubuntu. And Directus is able to communicate with MySQL just fine because the installation completed successfully and I can see all the newly created tables in the DB.

I thought maybe my network is causing issues, so I tried installing Directus on the same server which hosts MySQL and got exactly the same error.

**Comment by rijkvanzanten** [comment]
@jsthind Can you try setting the `LOG_LEVEL` environment variable to `trace` and see where it hangs?

**Comment by jsthind** [comment]
Ok, will get the details for you tomorrow from the office. I am in a different timezone and it is 10:00 PM here.

Let me know if you need any other details.

**Comment by rijkvanzanten** [comment]
Anything that'd help me reproduce this on my end would be very welcome 🙂

**Comment by Fusseldieb** [comment]
@jsthind Did you get any further on this?

**Comment by jsthind** [comment]
I added the trace log as suggested and found that the code blocked at the below query:

```sql
SELECT
	`c`.`TABLE_NAME`,
	`c`.`COLUMN_NAME`,
	`c`.`COLUMN_DEFAULT`,
	`c`.`DATA_TYPE`,
	`c`.`CHARACTER_MAXIMUM_LENGTH`,
	`c`.`IS_NULLABLE`,
	`c`.`COLUMN_KEY`,
	`c`.`EXTRA`,
	`c`.`COLLATION_NAME`,
	`c`.`COLUMN_COMMENT`,
	`c`.`NUMERIC_PRECISION`,
	`c`.`NUMERIC_SCALE`,
	`fk`.`REFERENCED_TABLE_NAME`,
	`fk`.`REFERENCED_COLUMN_NAME`,
	`fk`.`CONSTRAINT_NAME`,
	`rc`.`UPDATE_RULE`,
	`rc`.`DELETE_RULE`,
	`rc`.`MATCH_OPTION`
FROM
	`INFORMATION_SCHEMA`.`COLUMNS` AS `c`
	LEFT JOIN `INFORMATION_SCHEMA`.`KEY_COLUMN_USAGE` AS `fk` ON `c`.`TABLE_NAME` = `fk`.`TABLE_NAME`
		AND `fk`.`COLUMN_NAME` = `c`.`COLUMN_NAME`
		AND `fk`.`CONSTRAINT_SCHEMA` = `c`.`TABLE_SCHEMA`
	LEFT JOIN `INFORMATION_SCHEMA`.`REFERENTIAL_CONSTRAINTS` AS `rc` ON `rc`.`TABLE_NAME` = `fk`.`TABLE_NAME`
		AND `rc`.`CONSTRAINT_NAME` = `fk`.`CONSTRAINT_NAME`
		AND `rc`.`CONSTRAINT_SCHEMA` = `fk`.`CONSTRAINT_SCHEMA`
WHERE
	`c`.`TABLE_SCHEMA` = 'directus'
```


I googled a bit and found that slow information_schema in MySQL is a common problem. And the fact that my server has approx 450 databases and thousands of tables only make this query worse.

Not sure if you can implement a workaround for this.

**Comment by rijkvanzanten** [comment]
> Not sure if you can implement a workaround for this.

No idea! I don't think we can pull all the required information from another source 🤔

**Comment by nachogarcia** [comment]
I'm experiencing the same issue, with /server/ping taking seconds.

Maybe we can set `innodb_stats_on_metadata=0`

https://www.percona.com/blog/2011/12/23/solving-information_schema-slowness/

**Reply by rijkvanzanten** [reply]
Where does one set that?

**Reply by nachogarcia** [reply]
It's a [system variable](https://dev.mysql.com/doc/refman/8.0/en/innodb-parameters.html#sysvar_innodb_stats_on_metadata), nothing we can do at directus level I think.

It can be added by using `SET GLOBAL innodb_stats_on_metadata=0;` and adding to the .cnf

```
[mysqld]
innodb_stats_on_metadata = 0
```

**Reply by nickrum** [reply]
Seems like this is already the default for MySQL 5.7 (see [here](https://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html#sysvar_innodb_stats_on_metadata)).

According to [this](https://mysqlserverteam.com/mysql-8-0-scaling-and-performance-of-information_schema/) the performance of `INFORMATION_SCHEMA` queries has greatly improved with MySQL 8. For those stuck with version 5.7 or MariaDB, maybe we could optimize the relevant queries according to [this](https://dev.mysql.com/doc/refman/5.7/en/information-schema-optimization.html).

**Comment by lukas-schaetzle** [comment]
I'm also experiencing extremly long wait times for starting Directus when executing it on a shared host (I guess the problem is here that there are a lot of databases from different customers.) The problematic query is this one:

```
[57895.721ms]
      SELECT DISTINCT
        rc.TABLE_NAME AS 'table',
        kcu.COLUMN_NAME AS 'column',
        rc.REFERENCED_TABLE_NAME AS 'foreign_key_table',
        kcu.REFERENCED_COLUMN_NAME AS 'foreign_key_column',
        rc.CONSTRAINT_NAME AS 'constraint_name',
        rc.UPDATE_RULE AS on_update,
        rc.DELETE_RULE AS on_delete
      FROM
        information_schema.referential_constraints AS rc
      JOIN information_schema.key_column_usage AS kcu ON
        rc.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME
      WHERE
        rc.CONSTRAINT_SCHEMA = <database_name>;
```
Because of that same query the login also often takes very long. However, if I add anothere WHERE condition for the joined table, the query takes under 1 second.

```
<...>
      WHERE
        rc.CONSTRAINT_SCHEMA = <database_name> AND
        kcu.CONSTRAINT_SCHEMA = <database_name>;
```

**Reply by rijkvanzanten** [reply]
That looks awfully similar to the tweak @u12206050 just made in https://github.com/knex/knex-schema-inspector/pull/123. @Tummerhore Could you do me a favor and see if you get the same speed improvement by using 

```
      JOIN information_schema.key_column_usage AS kcu ON
        rc.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME AND
        kcu.CONSTRAINT_SCHEMA = <database_name>;
      WHERE
        rc.CONSTRAINT_SCHEMA = <database_name>;
```

instead of doing it under WHERE?

**Reply by lukas-schaetzle** [reply]
Ah, perfect. Didn't notice that pull request. That seems to have the same effect (at least on my server) :)

**Reply by rijkvanzanten** [reply]
> Didn't notice that pull request

No hard feelings, it was a PR for another bug in a separate repo 😄

**Reply by bobbyphilip** [reply]
@rijkvanzanten Was there a change in Directus for this, or was the only change using the updated knex-schema-inspector (v2.0.4 I think)
I am on Directus 9.17 and my reading of the commit history says that the knex-schema-inspector change made it in 9.16

**Reply by rijkvanzanten** [reply]
@bobbyphilip Schema inspection happens in the schema inspector library, so yeah, updating that is sufficient for the improvements in directus 👍🏻
